### PR TITLE
Migration from CHT 3.x on docker-compose to CHT 4.x on k3s

### PR DESCRIPTION
**Pull request by henokgetachew**
# Description

* How to migrate a CHT 3.x deployment that's on docker-compose to a single node k3s instance
* How to migrate a CHT 3.x deployment that's on docker-compose to a 3-node clustered couchdb on k3s.

I have made them separate files because it became complicated as a one file with if/else conditions. It makes them more readable this way. 

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+nodes:
+  node-1: "couch01"
+  node-2: "couch02"
+  node-3: "couch03"
+
+couchdb_data:
+  preExistingDataAvailable: "true"
+  dataPathOnDiskForCouchDB: "data"
+  partition: "0"
```
Comment (dianabarsan) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:112]: So here they're supposed to use the data storage that they just copied right? Is that mentioned somewhere?

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    user@node3-hostname:/srv/couchdb3/data/shards/
+
+# Then copy hidden shards
+sudo rsync -avz --progress --partial --partial-dir=/tmp/rsync-partial \
+    /srv/couchdb1/data/.shards/40000000-5fffffff \
+    /srv/couchdb1/data/.shards/a0000000-bfffffff \
+    user@node3-hostname:/srv/couchdb3/data/.shards/
+
+# Touch the .shards to ensure they're newer
+ssh user@node3-hostname "sudo find /srv/couchdb3/data/.shards -type f -exec touch {} +"
```
Comment (dianabarsan) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:203]: Newer than what? Why is this needed?

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  dataPathOnDiskForCouchDB: "data"
+  partition: "0"
+
+local_storage:
+  preExistingDiskPath-1: "/srv/couchdb1"
+```
+
+7. Deploy to K3s
+```shell
+cd cht-core/scripts/deploy
```
Comment (dianabarsan) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md:110]: CHT-Deploy just bursts out of nowhere :D 
I think we should link to what this is.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice. I'm not sure we need two pages, since there is a lot of repetition. Or is there a way we can get the common parts together and just reference them? I think there was one instance of this ... 
Ah: https://github.com/medic/cht-docs/blob/8101da95043af85d5c17099376ac03f65d3871ff/content/en/building/reference/_partial_cht_api.md?plain=1#L4

This page is reused in some other page like: https://github.com/search?q=repo%3Amedic%2Fcht-docs+_partial_cht_api&type=code

So I think all common elements, in both PRs, can be collected into partials and shared across.

**Review by mrjones-plip — CHANGES_REQUESTED** [review]
Great work here - thanks Henok!

I think using an shared `_include-file` is the way to go for content that overlaps per Diana's suggestion. Let me know if you need a hand!

I think the link titles are a bit confusing currently:

* 4.x
   * Data migration to 4.x
   * Docker Compose to K3s Clustered Migration
   * Docker Compose to K3s Single-Node Migration

Instead, I think we can clean them up to be more clear (possibly a sub-directory?  seems a bit too nested...we'll go flat for now!):

* 4.x 
   * Migration: 3.x to 4.x
   * Migration: 4.x Docker to 4.x k3s (Multi-node)
   * Migration: 4.x Docker to 4.x k3s (Single-node)

I've made suggested code changes for this, but can't edit the `Migration: 3.x to 4.x` document link in the PR, please make the change though.
 
As well, we should make each number a heading so it's clickable and appears in the right hand side TOC. I've made suggested code changes for this.

Variables should always be the same to reduce confusion (eg `username` vs `admin-user` vs `saved-user-login`). I've made suggested code changes for this, but may have missed some!

It's incredibly daunting have a YAML file and try and read through any changes that are needed. I recommend we:
* always list out the items that need changing
* always use the same syntax to highlight an item need changing using angle brackets: `<change-me>`
* Always use the same name for the same value in and out of the YAML(mentioned above)

**Review by henokgetachew — COMMENTED** [review]
Accepted most of the suggested changes. I'm now extracting the partial.

**Comment by henokgetachew** [comment]
FYI, I'm rebasing this PR against the main branch.

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
+---
+title: "Migration from Docker Compose CHT 3.x to 3-Node Clustered CHT 4.x on K3s"
+linkTitle: "Migration: 3.x Docker Compose to 4.x K3s (Multi-node)"
+weight: 1
+description: >
+  Guide to migrate existing data from CHT 3.x Docker Compose deployment to CHT 4.x clustered K3s deployment with 3 CouchDB nodes
+relatedContent: >
+---
+{{ < read-content file="hosting/4.x/_partial_migration_3x_docker_to_4x_k3s.md"  }}
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:9]: missing closing `>` and can't have leading spaces around `<` :

```suggestion
{{< read-content file="hosting/4.x/_partial_migration_3x_docker_to_4x_k3s.md"  >}}
```

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+# Create directories on secondary nodes
+
+```shell
+ssh <user>@<node2-hostname> "sudo mkdir -p /srv/couchdb2/data/shards /srv/couchdb2/data/.shards"
+ssh <user>@<node3-hostname> "sudo mkdir -p /srv/couchdb3/data/shards /srv/couchdb3/data/.shards"
+```
+
+## Create values.yaml for K3s Deployment
+{{ < read-content file="hosting/4.x/_partial_values_explanation.md"  }}
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:19]: missing closing `>` and can't have leading spaces around `<` :

```suggestion
{{< read-content file="hosting/4.x/_partial_values_explanation.md"  >}}
```

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
+---
+title: "Migration from 3.x Docker Compose to 4.x K3s (Single Node)"
+linkTitle: "Migration: 3.x Docker Compose to 4.x K3s (Single-Node)"
+weight: 1
+description: >
+  Guide on how to migrate existing data from CHT 3.x Docker Compose deployment to CHT 4.x single-node K3s deployment
+relatedContent: >
+---
+{{ < read-content file="hosting/4.x/_partial_migration_3x_docker_to_4x_k3s.md"  }}
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md:9]: missing closing `>` and can't have leading spaces around `<` :

```suggestion
{{< read-content file="hosting/4.x/_partial_migration_3x_docker_to_4x_k3s.md"  >}}
```

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+## Create values.yaml for K3s Deployment
+{{ < read-content file="hosting/4.x/_partial_values_explanation.md"  }}
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md:12]: missing closing `>` and can't have leading spaces around `<` :

```suggestion
{{< read-content file="hosting/4.x/_partial_values_explanation.md"  >}}
```

File: content/en/hosting/4.x/_partial_migration_3x_docker_to_4x_k3s.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+cd ~/couchdb-migration/
+curl -s -o ./docker-compose.yml https://raw.githubusercontent.com/medic/couchdb-migration/main/docker-compose.yml
+docker compose up
+```
+
+## Set Up Environment Variables
+```shell
+# Replace with your actual CouchDB URL from the Docker Compose setup
+export COUCH_URL=http://<your-admin-user>:<your-admin-password>@<couchdb-host>:5984
+```
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/_partial_migration_3x_docker_to_4x_k3s.md:25]: We should standardize on using the same `<variable-name>`  for username and password.  As well, let's pull the code comment out to be text that's easier to read:

``````suggestion
## Set Up Environment Variables

Be sure to replace both `<admin-user>` and `<password>` with your actual username and password.  As well, update `<couchdb-host>` to the CouchDB URL from the Docker Compose setup:

```shell
export COUCH_URL=http://<admin-user>:<password>@<couchdb-host>:5984
```
``````

File: content/en/hosting/4.x/_partial_values_explanation.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+hide_summary: true
+---
+
+* `<your-namespace>` (_two occurrences_)
+* `<version>` - 4.x version you're upgrading to
+* `<password>` - retrieved from `get-env` call above
+* `<secret>` - retrieved from `get-env` call above
+* `<admin_user>` - needs to be the same as used in 3.x - likely `medic`
+* `<uuid>` - retrieved from `get-env` call above
+* `<url>` - your DNS goes here (your-dns-address.org)
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/_partial_values_explanation.md:12]:
```suggestion
---

Be sure to update the following values in your YAML file:

* `<your-namespace>` (_two occurrences_)
* `<version>` - 4.x version you're upgrading to
* `<password>` - retrieved from `get-env` call above
* `<secret>` - retrieved from `get-env` call above
* `<admin_user>` - needs to be the same as used in 3.x - likely `medic`
* `<uuid>` - retrieved from `get-env` call above
* `<url>` - the URL of your production instance goes here (eg `www.example.org`)
* `<path-to-tls>` - path to TLS files on disk
```

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+
+```yaml
+project_name: "your-namespace"
+namespace: "your-namespace"
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:23]:
```suggestion
project_name: "<your-namespace>"
namespace: "<your-namespace>"
```

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+```yaml
+project_name: "your-namespace"
+namespace: "your-namespace"
```
Comment (mrjones-plip) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md:16]:
```suggestion
project_name: "<your-namespace>"
namespace: "<your-namespace>"
```

**Review by mrjones-plip — CHANGES_REQUESTED** [review]
Coming along!  Nice work here.

See below for requested changes.

It's also not clear what the "Data migration to 4.x" doc is for.  Is that for both docker and k3s?  Is that more generically for how to migrate from 3.x to 4.x using the [couchdb-migration](https://github.com/medic/couchdb-migration) tool? I think we should update the link title to be more accurate.  Maybe that's:

`Data migration to 4.x` -> `Migration: Docker 3.x to Docker 4.x` - but I'm not sure!

As well, we need to clarify on the page itself as it just says:

> Guide to migrate existent data from CHT 3.x to CHT 4.x

But doesn't qualify it any further.  The two new pages do a great job of this!

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+```shell
+ssh <user>@<node2-hostname> "sudo mkdir -p /srv/couchdb2/data/shards /srv/couchdb2/data/.shards"
+ssh <user>@<node3-hostname> "sudo mkdir -p /srv/couchdb3/data/shards /srv/couchdb3/data/.shards"
+```
+
+## Create values.yaml for K3s Deployment
+{{ < read-content file="hosting/4.x/_partial_values_explanation.md"  }}
+
+```yaml
+project_name: "your-namespace"
```
Comment (dianabarsan) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:22]: It could be tough to keep the documentation _and_ the helm repo in sync. Can we, instead of inlining the whole contents of the values file somehow link to the current version of it?

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md (hunk 1 above)
Comment (dianabarsan) [reply content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:60]: Is this done? I can't find a reference yet.

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  preExistingDataAvailable: "true"
+  dataPathOnDiskForCouchDB: "data"
+  partition: "0"
+
+local_storage:
+  preExistingDiskPath-1: "/srv/couchdb1"
+```
+
+## Deploy to K3s
+```shell
```
Comment (dianabarsan) [inline content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md:55]: Please add a comment like: 
```
Use cht-deploy (link) to deploy CHT-Core to K8s. 
```

**Review by dianabarsan — COMMENTED** [review]
Cool! Left some minor comments :)

File: content/en/hosting/4.x/_partial_values_explanation.md (hunk 9 above)
Comment (henokgetachew) [reply content/en/hosting/4.x/_partial_values_explanation.md]: I have removed the www and used just example.com.

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md (hunk 13 above)
Comment (henokgetachew) [reply content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md:55]: I've added this in the partial to avoid duplication. Right before the prepare node section.

>We are also going to utilize the `cht-deploy` script from the [cht-core](https://github.com/medic/cht-core) repo. If you don't already have that, clone it.

Still think it should come here?

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md (hunk 1 above)
Comment (henokgetachew) [reply content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:60]: Ah. Pushing an update now.

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md (hunk 12 above)
Comment (henokgetachew) [reply content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md]: I have somehow struggled with this too. The original [values.yaml](https://github.com/medic/helm-charts/blob/main/charts/cht-chart-4x/values.yaml) in the helm charts is much larger and robust as it accounts for all the different scenarios. I originally thought that the comments on the yaml would be enough to guide people to the correct final configuration of what should be removed or what should be included. However, that hasn't been the case. That's why I decided to add example config of the various scenarios in the docs. Although not ideal, we'd have to remember to keep the docs in sync when we add/remove features from the helm-charts.

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md (hunk 1 above)
Comment (henokgetachew) [reply content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-multi.md:60]: Done.

**Comment by henokgetachew** [comment]
>It's also not clear what the "Data migration to 4.x" doc is for. Is that for both docker and k3s? Is that more generically for how to migrate from 3.x to 4.x using the [couchdb-migration](https://github.com/medic/couchdb-migration) tool? I think we should update the link title to be more accurate.

@mrjones-plip I have renamed the file and changed the description as well.

**Review by mrjones-plip — CHANGES_REQUESTED** [review]
Very close! I noticed a few discrepancies in how we name inline variables that we want the user to replace - I've made suggestions to make these all homogeneous.

I see you [made some changes](https://github.com/medic/cht-docs/pull/1723/commits/6fdfd3225f4af10b58cb78db0495f34a16850c07) to the frontmatter of the internal Medic EKS guide.  This is not what I was [suggesting](https://github.com/medic/cht-docs/pull/1723#pullrequestreview-2459480376).  Instead, I suggest we change the name of "Migration from CHT 3.x to CHT 4.x" [document here](https://github.com/medic/cht-docs/blob/main/content/en/hosting/4.x/data-migration.md) - this document is not clear what it applies to.  However!  After reviewing this PR for a third time, I think we should just create a new directory to simplify the left hand navigation and page titles.  ALL The documents are 3.x to 4.x and they're ALL from docker hosting ;)  I propose we take this:

image

and create a "Migration guides" directory and rename the links like so:

image

File: content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md (hunk 13 above)
Comment (henokgetachew) [reply content/en/hosting/4.x/data-migration-3x-docker-to-4x-k3s-single.md]: Good point, I've added a note right above it.

**Review by henokgetachew — COMMENTED** [review]
Good idea. I have re-organized docs into a migration dir.

**Comment by henokgetachew** [comment]
@mrjones-plip I have also updated the links in other docs referencing the migration page. The build is passing and I think iit should work fine.  Let me know if anything is broken or feel free to also push directly to the branch if you think there should be a different folder structure.

**Comment by mrjones-plip** [comment]
cool - I'll take a look later today and get back to you!

**Review by mrjones-plip — APPROVED** [review]
Thanks so much!  Made a minor fix in b742888 , but this is good to go. w00t \o/

**Review by dianabarsan — APPROVED** [review]
